api/kv.go

```diff
@@ -0,0 +1,328 @@
+	// Keys are stringified ints, e.g. "3"
+	Versions map[string]VersionMetadata `mapstructure:"versions"`
+}
+
+type VersionMetadata struct {
+	Version      int       `mapstructure:"version"`
+	CreatedTime  time.Time `mapstructure:"created_time"`
+	DeletionTime time.Time `mapstructure:"deletion_time"`
+	Destroyed    bool      `mapstructure:"destroyed"`
+	// There is currently no version-specific custom metadata.
+	// This field is just a copy of what's in the CustomMetadata field
+	// for the full KVMetadata of the secret.
+	CustomMetadata map[string]string `mapstructure:"custom_metadata"`
+}
+
+func (c *Client) KVv1(mountPath string) *KVClient {
+	return &KVClient{c: c, mountPath: mountPath, version: 1}
+}
+
+func (c *Client) KVv2(mountPath string) *KVClient {
+	return &KVClient{c: c, mountPath: mountPath, version: 2}
+}
+
+func (kv *KVClient) Read(ctx context.Context, secretPath string) (*KVSecret, error) {
+	pathToRead, err := kv.getFullPath(secretPath)
+	if err != nil {
+		return nil, fmt.Errorf("error assembling full path to KV secret: %v", err)
+	}
+
+	secret, err := kv.c.Logical().ReadWithContext(ctx, pathToRead)
+	if err != nil {
+		return nil, fmt.Errorf("error encountered while reading secret at %s: %v", pathToRead, err)
+	}
+	if secret == nil || secret.Data == nil {
+		return nil, fmt.Errorf("no secret found at %s", pathToRead)
+	}
+
+	return extractDataAndVersionMetadata(secret, kv.version)
+}
+
+// ReadVersion returns the data and metadata for a specific version of the
+// given secret. If that version has been deleted, the Data field on the
+// returned secret will be nil, and the Metadata field will contain the deletion time.
+func (kv *KVClient) ReadVersion(ctx context.Context, secretPath string, version int) (*KVSecret, error) {
+	pathToRead, err := kv.getFullPath(secretPath)
+	if err != nil {
+		return nil, fmt.Errorf("error assembling full path to KV secret: %v", err)
+	}
+
+	queryParams := map[string][]string{"version": {strconv.Itoa(version)}}
+	secret, err := kv.c.Logical().ReadWithDataWithContext(ctx, pathToRead, queryParams)
+	if err != nil {
+		return nil, err
+	}
+	if secret == nil || secret.Data == nil {
+		return nil, fmt.Errorf("no secret version found at %s", pathToRead)
+	}
+
+	return extractDataAndVersionMetadata(secret, kv.version)
+}
+
+func (kv *KVClient) ReadMetadata(ctx context.Context, secretPath string) (*KVMetadata, error) {
+	pathToRead, err := kv.getFullMetadataPath(secretPath)
+	if err != nil {
+		return nil, fmt.Errorf("error assembling full path to KV secret's metadata: %v", err)
+	}
+
+	secret, err := kv.c.Logical().ReadWithContext(ctx, pathToRead)
+	if err != nil {
+		return nil, err
+	}
+	if secret == nil || secret.Data == nil {
+		return nil, fmt.Errorf("no secret metadata found at %s", pathToRead)
+	}
+
+	return extractFullMetadata(secret)
+}
+
+func (kv *KVClient) Write(ctx context.Context, secretPath string, data map[string]interface{}) (*KVSecret, error) {
```

Comment:
Would it be a bit more idiomatic & consistent with CLI to go with Get/Put method names instead of Read/Write?

Reply:
Good question! I could go either way on this one. 

I considered Get/Put first but thought maybe it might be good to stick with simple Read/Write since app developers (who are just trying to grab a secret from this service called Vault that their ops team told them to use) may not be familiar with the CLI and therefore might not necessarily think to look for "Put" when they want to Create a resource for the first time (there's no Post after all). And "Users want to easily know how to read and write secrets" is the phrasing I keep hearing the most, at least from the product folks.

But then on the other hand, Get/Put could indeed be more appropriate if we implemented the Patch() method. And KV Patch and KV Metadata Patch are functionality we eventually need to support here, after all. 🤔

@VinnyHC Do you have an opinion here?

Reply:
No strong opinion here. Typically I associate `Put` with upserting information and being idempotent, which if I understand this correctly, this is not, since a new version is created on each success. On the other side, I think of `Read` and `Write` as binary options, and when you have more operations use a verb like `Create` as in Create, Read, Update, Delete, to be more specific.

Reply:
Ended up going with the Get/Put naming since Read/Write is too binary/restrictive, but Create/Update gets too confusing with the various differences between Vault's version of PUT and PATCH. Felt like it made more sense to stick with the naming scheme Vault already has come up with, so that users who are confused by the Put behavior can at least look at the kv put documentation pages to understand how it all works. 
Thanks for the suggestion!